### ISSUE TITLE
NooBaa CR gets updated multiple times per second / API spam

### ISSUE THREAD

**Issue by lallinger-arbeit** [noobaa/noobaa-operator#293]
Hello everybody,
after upgrading from noobaa-core 5.2.9 to 5.3.1 and from noobaa-operator 2.0.9 to 2.1.1 we are facing a problem where the noobaa CR changes multiple times per second resulting in frequent calls to the kubernetes API. Every time the noobaa CR changes only the "lastTransitionTime" and the "resourceVersion" fields are updated, nothing else.

This can easily be watched using ```oc get noobaa -w"```.

For example install noobaa-core 5.2.9 and noobaa-operator 2.0.9 using the CLI:
```noobaa install -n noobaa-old --noobaa-image='noobaa/noobaa-core:5.2.9' --operator-image='noobaa/noobaa-operator:2.0.9'```
When the installation has finished successfully watch the noobaa CR:
```oc get noobaa -n noobaa-old -w```
This returns one entry and the next time the CR changes/gets reconciled takes several minutes (as should be).

In contrary to this if you install the newest version of noobaa-core and noobaa-operator:
```noobaa install -n noobaa-new --noobaa-image='noobaa/noobaa-core:5.3.1' --operator-image='noobaa/noobaa-operator:2.1.1'```
And then watch the noobaa CR after successfull installation:
```oc get noobaa -n noobaa-new -w```
You can see, that the CR changes multiple times in a short period of time.

Getting the yaml of the CR a few seconds apart shows, that the CR changes multiple times a second. For example, i saved the yaml of the CR with ```oc get noobaa noobaa -n noobaa-new -o yaml > 1.yaml``` and a few seconds later again using ```oc get noobaa noobaa -n noobaa-new -o yaml > 2.yaml```. Comparing these two yamls you see only the field "lastTransitionTime" (which appears 3 times in the CR) and the "resourceVersion" changed. Calculating the difference in the "lastTransitionTime" and the "resourceVersion" shows that the CR changed approximately 4 times per second!

1.yaml:
```apiVersion: noobaa.io/v1alpha1
kind: NooBaa
metadata:
  creationTimestamp: "2020-04-20T16:16:22Z"
  generation: 1
  labels:
    app: noobaa
  name: noobaa
  namespace: noobaa-new
  resourceVersion: "69654985"
  selfLink: /apis/noobaa.io/v1alpha1/namespaces/noobaa-new/noobaas/noobaa
  uid: 46af0125-8322-11ea-ac51-000d3a22f627
spec:
  dbImage: centos/mongodb-36-centos7
  image: noobaa/noobaa-core:5.3.1
status:
  accounts:
    admin:
      secretRef:
        name: noobaa-admin
        namespace: noobaa-new
  actualImage: noobaa/noobaa-core:5.3.1
  conditions:
  - lastHeartbeatTime: "2020-04-20T16:16:36Z"
    lastTransitionTime: "2020-04-28T06:30:25Z"
    message: noobaa operator completed reconcile - system is ready
    reason: SystemPhaseReady
    status: "True"
    type: Available
  - lastHeartbeatTime: "2020-04-20T16:16:36Z"
    lastTransitionTime: "2020-04-28T06:30:25Z"
    message: noobaa operator completed reconcile - system is ready
    reason: SystemPhaseReady
    status: "False"
    type: Progressing
  - lastHeartbeatTime: "2020-04-20T16:16:36Z"
    lastTransitionTime: "2020-04-20T16:16:36Z"
    message: noobaa operator completed reconcile - system is ready
    reason: SystemPhaseReady
    status: "False"
    type: Degraded
  - lastHeartbeatTime: "2020-04-20T16:16:36Z"
    lastTransitionTime: "2020-04-28T06:30:25Z"
    message: noobaa operator completed reconcile - system is ready
    reason: SystemPhaseReady
    status: "True"
    type: Upgradeable
  endpoints:
    readyCount: 1
    virtualHosts:
    - s3.noobaa-new.svc
  observedGeneration: 1
  phase: Ready
  readme: "\n\n\tWelcome to NooBaa!\n\t-----------------\n\tNooBaa Core Version:     5.3.1-80055c5\n\tNooBaa
    Operator Version: 2.1.1\n\n\tLets get started:\n\n\t1. Connect to Management console:\n\n\t\tRead
    your mgmt console login information (email & password) from secret: \"noobaa-admin\".\n\n\t\t\tkubectl
    get secret noobaa-admin -n noobaa-new -o json | jq '.data|map_values(@base64d)'\n\n\t\tOpen
    the management console service - take External IP/DNS or Node Port or use port
    forwarding:\n\n\t\t\tkubectl port-forward -n noobaa-new service/noobaa-mgmt 11443:443
    &\n\t\t\topen https://localhost:11443\n\n\t2. Test S3 client:\n\n\t\tkubectl port-forward
    -n noobaa-new service/s3 10443:443 &\n\t\tNOOBAA_ACCESS_KEY=$(kubectl get secret
    noobaa-admin -n noobaa-new -o json | jq -r '.data.AWS_ACCESS_KEY_ID|@base64d')\n\t\tNOOBAA_SECRET_KEY=$(kubectl
    get secret noobaa-admin -n noobaa-new -o json | jq -r '.data.AWS_SECRET_ACCESS_KEY|@base64d')\n\t\talias
    s3='AWS_ACCESS_KEY_ID=$NOOBAA_ACCESS_KEY AWS_SECRET_ACCESS_KEY=$NOOBAA_SECRET_KEY
    aws --endpoint https://localhost:10443 --no-verify-ssl s3'\n\t\ts3 ls\n\n"
  services:
    serviceMgmt:
      externalDNS:
      - https://noobaa-mgmt-noobaa-new.customdomain.com
      internalDNS:
      - https://noobaa-mgmt.noobaa-new.svc:443
      internalIP:
      - https://172.30.149.9:443
      nodePorts:
      - https://100.66.29.4:31750
      podPorts:
      - https://10.131.0.101:8443
    serviceS3:
      externalDNS:
      - https://s3-noobaa-new.customdomain.com
      internalDNS:
      - https://s3.noobaa-new.svc:443
      internalIP:
      - https://172.30.6.184:443
      nodePorts:
      - https://100.66.29.4:31753
      podPorts:
      - https://10.131.0.102:6443
```

2.yaml:
```apiVersion: noobaa.io/v1alpha1
kind: NooBaa
metadata:
  creationTimestamp: "2020-04-20T16:16:22Z"
  generation: 1
  labels:
    app: noobaa
  name: noobaa
  namespace: noobaa-new
  resourceVersion: "69655062"
  selfLink: /apis/noobaa.io/v1alpha1/namespaces/noobaa-new/noobaas/noobaa
  uid: 46af0125-8322-11ea-ac51-000d3a22f627
spec:
  dbImage: centos/mongodb-36-centos7
  image: noobaa/noobaa-core:5.3.1
status:
  accounts:
    admin:
      secretRef:
        name: noobaa-admin
        namespace: noobaa-new
  actualImage: noobaa/noobaa-core:5.3.1
  conditions:
  - lastHeartbeatTime: "2020-04-20T16:16:36Z"
    lastTransitionTime: "2020-04-28T06:30:43Z"
    message: noobaa operator completed reconcile - system is ready
    reason: SystemPhaseReady
    status: "True"
    type: Available
  - lastHeartbeatTime: "2020-04-20T16:16:36Z"
    lastTransitionTime: "2020-04-28T06:30:43Z"
    message: noobaa operator completed reconcile - system is ready
    reason: SystemPhaseReady
    status: "False"
    type: Progressing
  - lastHeartbeatTime: "2020-04-20T16:16:36Z"
    lastTransitionTime: "2020-04-20T16:16:36Z"
    message: noobaa operator completed reconcile - system is ready
    reason: SystemPhaseReady
    status: "False"
    type: Degraded
  - lastHeartbeatTime: "2020-04-20T16:16:36Z"
    lastTransitionTime: "2020-04-28T06:30:43Z"
    message: noobaa operator completed reconcile - system is ready
    reason: SystemPhaseReady
    status: "True"
    type: Upgradeable
  endpoints:
    readyCount: 1
    virtualHosts:
    - s3.noobaa-new.svc
  observedGeneration: 1
  phase: Ready
  readme: "\n\n\tWelcome to NooBaa!\n\t-----------------\n\tNooBaa Core Version:     5.3.1-80055c5\n\tNooBaa
    Operator Version: 2.1.1\n\n\tLets get started:\n\n\t1. Connect to Management console:\n\n\t\tRead
    your mgmt console login information (email & password) from secret: \"noobaa-admin\".\n\n\t\t\tkubectl
    get secret noobaa-admin -n noobaa-new -o json | jq '.data|map_values(@base64d)'\n\n\t\tOpen
    the management console service - take External IP/DNS or Node Port or use port
    forwarding:\n\n\t\t\tkubectl port-forward -n noobaa-newservice/noobaa-mgmt 11443:443
    &\n\t\t\topen https://localhost:11443\n\n\t2. Test S3 client:\n\n\t\tkubectl port-forward
    -n noobaa-newservice/s3 10443:443 &\n\t\tNOOBAA_ACCESS_KEY=$(kubectl get secret
    noobaa-admin -n noobaa-new -o json | jq -r '.data.AWS_ACCESS_KEY_ID|@base64d')\n\t\tNOOBAA_SECRET_KEY=$(kubectl
    get secret noobaa-admin -n noobaa-new -o json | jq -r '.data.AWS_SECRET_ACCESS_KEY|@base64d')\n\t\talias
    s3='AWS_ACCESS_KEY_ID=$NOOBAA_ACCESS_KEY AWS_SECRET_ACCESS_KEY=$NOOBAA_SECRET_KEY
    aws --endpoint https://localhost:10443 --no-verify-ssl s3'\n\t\ts3 ls\n\n"
  services:
    serviceMgmt:
      externalDNS:
      - https://noobaa-mgmt-noobaa-new.customdomain.com
      internalDNS:
      - https://noobaa-mgmt.noobaa-new.svc:443
      internalIP:
      - https://172.30.149.9:443
      nodePorts:
      - https://100.66.29.4:31750
      podPorts:
      - https://10.131.0.101:8443
    serviceS3:
      externalDNS:
      - https://s3-noobaa-new.customdomain.com
      internalDNS:
      - https://s3.noobaa-new.svc:443
      internalIP:
      - https://172.30.6.184:443
      nodePorts:
      - https://100.66.29.4:31753
      podPorts:
      - https://10.131.0.102:6443
```

This frequent changes of the CR look like a bug introduced in the latest version and should be fixed.

**Comment by nimrod-becker** [comment]
Hi @lallinger-arbeit thanks for reporting this!
Might be related, we found a similar issue of an infinite update cycle (see https://github.com/noobaa/noobaa-core/pull/5983) which was caused by the endpoint updates.

This will be released in the 5.4 version of noobaa-core, if you can use the operator master, it points to an RC of that version. 

I'm not sure this is exactly the same, but it looks awfully close.

**Comment by lallinger-arbeit** [comment]
Just tried it with the current master state of noobaa-core and noobaa-operator. When using ```oc get noobaa -w``` it seems to update less often (only every few seconds vs. 2 times every second before). Just looking at the "resourceVersion" and the "lastTransitionTime" it still updates multiple times per second (i'm not sure if the "resourceVersion" really increases only by one every time an update happens).
I still think this is quite often because nothing in the CR is changing.
These updates won't be a problem when only using noobaa, but we want to deploy noobaa via the operatorhub and ArgoCD. Because ArgoCD watches for changes in objects it deployed this doubles the calls to the kubernetes API, which can result in crashes of the API pods.

Are these updates of the CR really needed if nothing changes, except the "lastTransitionTime"?

**Comment by guymguym** [comment]
@lallinger-arbeit No - off course these vain updates are not needed, we need to catch the issuer of those updates - @dannyzaken did you catch it already?

**Comment by lallinger-arbeit** [comment]
Happy to see this getting handled!

**Comment by lallinger-arbeit** [comment]
this seems to be mostly fixed in 207b8f5dbd0d1ee73363d10541a626a6a5eeb661 , resource only gets update about once every other second.
